File: Data Structures & Algorithms/products-of-array-discluding-self/submission-2.go
func productExceptSelf(nums []int) []int {
    n := len(nums)
    prefixProduct, postfixProduct := make([]int, n), make([]int, n)
    prefixProduct[0], postfixProduct[n-1] = 1, 1
    for i := n-2; i >= 0; i-- {
        postfixProduct[i] = postfixProduct[i+1] * nums[i+1]
    }
    for i := 1; i < n; i++ {
        prefixProduct[i] = prefixProduct[i-1] * nums[i-1]
    }
    res := make([]int, n)
    for i := range nums {
       res[i] =  prefixProduct[i] * postfixProduct[i]
    }
    return res
}

File: Data Structures & Algorithms/pascals-triangle/submission-2.py
class Solution:
    def generate(self, numRows: int) -> List[List[int]]:
        res = []
        for i in range(numRows):
            row = [1]*(i+1)
            # Only compute j = 1..i-1 because row[0] and row[i] are always 1 in Pascal's triangle
            for j in range(1, i):
                # Use Pascal's recurrence from the previous row to avoid integer division and make the update simpler and less error-prone.
                row[j] = res[i-1][j-1] + res[i-1][j]
            res.append(row)
        return res